idx/memory/find_cache.go

```diff
@@ -0,0 +1,190 @@
+package memory
+
+import (
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/grafana/metrictank/stats"
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/raintank/schema"
+	log "github.com/sirupsen/logrus"
+)
+
+var (
+	// metric idx.memory.find-cache.hit is a counter of findCache hits
+	findCacheHit = stats.NewCounterRate32("idx.memory.find-cache.hit")
+	// metric idx.memory.find-cache.miss is a counter of findCache misses
+	findCacheMiss = stats.NewCounterRate32("idx.memory.find-cache.miss")
+
+	findCacheSize            = 1000
+	findCacheInvalidateQueue = 100
+	findCacheBackoff         = time.Minute
+)
+
+type FindCache struct {
+	cache map[uint32]*lru.Cache
+	size  int
+	sync.RWMutex
+	newSeries map[uint32]chan struct{}
+	backoff   map[uint32]time.Time
+}
+
+func NewFindCache(size int) *FindCache {
+	fc := &FindCache{
+		cache:     make(map[uint32]*lru.Cache),
+		size:      size,
+		newSeries: make(map[uint32]chan struct{}),
+		backoff:   make(map[uint32]time.Time),
+	}
+	return fc
+}
+
+func (c *FindCache) Get(orgId uint32, pattern string) ([]*Node, bool) {
+	c.RLock()
+	cache, ok := c.cache[orgId]
+	c.RUnlock()
+	if !ok {
+		findCacheMiss.Inc()
+		return nil, false
+	}
+	nodes, ok := cache.Get(pattern)
```

Comment:
Is there a chance that `cache` could have already been deleted before attempting this call in some fringe situations?

Reply:
It is possible that one thread calls `cache, ok := c.cache[orgId]` then releases the lock, then before `nodes, ok := cache.Get(pattern)` is executed another thread gets a write lock and calls `delete(c.cache, orgId)`.  But it doesnt really matter.  But this is still safe as the items in c.cache[orgId] are pointers.  The end result is that calls to `findCache.Get()` will return results based on the content of cache when the Rlock() was acquired an not on the contents of the cache at the specific time that `nodes, ok := cache.Get(pattern)` is executed.